### PR TITLE
Validation of identity with Gatekeeper

### LINKED ISSUE #72: Add Validation Web hook for Azure Identity is valid
Adding a validation web hook 
- AzureIdentity is created
- AzureIdentityBinding is created

### PR DESCRIPTION
**Pull request by surenderssm**
**Reason for Change**:
This will help validate various CRDs and the azure resources used in aad-pod-identity.
Currently validation of format of the User assigned MSI is supported.

**Issue Fixed**:
Fixes
#72 
https://github.com/Azure/aad-pod-identity/issues/289

**Notes for Reviewers**:
Given constraint will be applicable for all requests on  Kind `AzureIdentity` in apiGroup `aadpodidentity.k8s.io`
We can add similar constraint across other resources in upcoming PR.

### CODE REVIEW

File: validation/gatekeeper/azureidentityformat_template.yaml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        kind: azureidentityformat
+        listKind: azureidentityformatList
+        plural: azureidentityformat
+        singular: azureidentityformat
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package azureidentityformat
+        violation[{"msg": msg}] {
+         input.review.object.kind == "AzureIdentity"
```
Comment (ritazh) [inline validation/gatekeeper/azureidentityformat_template.yaml]: Consider using `input.review.kind.kind` instead

Comment (surenderssm) [reply validation/gatekeeper/azureidentityformat_template.yaml]: thank you @ritazh , have acted on the suggestion.

**Comment by surenderssm** [comment]
Copying the result from the constraint test added in e2e.
$ go test D:/lab/go/aad-pod-identity/test/e2e -count=1 -v -timeout 24h -ginkgo.progress
=== RUN   TestAADPodIdentity
Running Suite: AAD Pod Identity Suite
=====================================
Random Seed: 1567598979
Will run 1 of 19 specs

Setting up the test suite environment...
Creating directory  template/_output
System MSI enabled: false
$ kubectl apply -f template/_output/default-deployment.yaml
# Poll to check if nmi daemonset is ready...
$ kubectl get ds -ojson
# Poll to check if busybox daemonset is ready...
$ kubectl get ds -ojson
# busybox daemonset is not ready yet. Retrying in 3s...
$ kubectl get ds -ojson
$ kubectl get pods -ojson
$ kubectl exec busybox-s8ftj -- apk add iptables
# Tight-poll to check if iptable rules for aad-metadata exist on all nodes...
$ kubectl get pods -ojson
$ kubectl exec busybox-s8ftj -- iptables -t nat --check PREROUTING -j aad-metadata
$ kubectl exec busybox-s8ftj -- iptables -t nat -L aad-metadata
SSSSSSSSSSSSSSSSSS

[1] should pass the identity format validation via gatekeeper constraint
$ kubectl apply -f https://raw.githubusercontent.com/open-policy-agent/gatekeeper/master/deploy/gatekeeper.yaml
$ kubectl apply -f ../../validation/gatekeeper/azureidentityformat_template.yaml
constrainttemplate.templates.gatekeeper.sh/azureidentityformat created
$ kubectl apply -f ../../validation/gatekeeper/azureidentityformat_constraint.yaml
azureidentityformat.constraints.gatekeeper.sh/azureidentityformatconstraint created
$ kubectl apply -f template/aadpodidentity_test_invalid.yaml
Error from server ([denied by azureidentityformatconstraint] The identity resourceId '/subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.ManagedIdentity/userAssignedIdentities/myidentity' is invalid.It must be of the following format: '/subscriptions/<subid>/resourcegroups/<resourcegroup>/providers/Microsoft.ManagedIdentity/userAssignedIdentities/<name>'): error when creating "template/aadpodidentity_test_invalid.yaml": admission webhook "validation.gatekeeper.sh" denied the request: [denied by azureidentityformatconstraint] The identity resourceId '/subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.ManagedIdentity/userAssignedIdentities/myidentity' is invalid.It must be of the following format: '/subscriptions/<subid>/resourcegroups/<resourcegroup>/providers/Microsoft.ManagedIdentity/userAssignedIdentities/<name>'
$ kubectl apply -f template/aadpodidentity_test_valid.yaml
azureidentity.aadpodidentity.k8s.io/testidentityvalid created
$ kubectl delete -f ../../validation/gatekeeper/azureidentityformat_constraint.yaml
$ kubectl delete -f ../../validation/gatekeeper/azureidentityformat_template.yaml
$ kubectl delete -f https://raw.githubusercontent.com/open-policy-agent/gatekeeper/master/deploy/gatekeeper.yaml

Tearing down the test environment...
$ kubectl delete AzureIdentity,AzureIdentityBinding --all
# Tight-poll to check if the Azure Assigned Identity is deleted...
$ kubectl get AzureAssignedIdentity -ojson
$ kubectl delete deploy -l app=identity-validator
# Tight-poll to check if pods are deleted...
$ kubectl get pods -ojson

------------------------------
+ [SLOW TEST:202.515 seconds]
Kubernetes cluster using aad-pod-identity
D:/lab/go/aad-pod-identity/test/e2e/aadpodidentity_test.go:111
  should pass the identity format validation via gatekeeper constraint
  D:/lab/go/aad-pod-identity/test/e2e/aadpodidentity_test.go:920
------------------------------

Tearing down the test suite environment...
$ kubectl delete -f template/_output/default-deployment.yaml --now --ignore-not-found
$ kubectl delete -f template/busyboxds.yaml
$ kubectl delete deploy --all
$ kubectl get nodes -ojson
# Removing system assigned identity from aks-nodepool1-33682749-0...

Ran 1 of 19 Specs in 301.373 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 18 Skipped
PASS | FOCUSED

File: docs/readmes/README.features.md [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# Features
+
+1. [Match Pods in the Namespace](README.namespaced.md)
+2. [System MSI cluster support](README.msi.md)
+3. [Non cluster credentials for MIC](README.noazure-json.md)
+4. [Application exception](README.app-exception.md)
+5. [Validation](README.validation.md)
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply docs/readmes/README.features.md]: Moved.

File: docs/readmes/README.validation.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
+# Azure Identity Validation
+
+## Introduction
+
+This will help validate various CRDs and the azure resources used in aad-pod-identity.
+Currently validation of format of the User assigned MSI is supported.
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply docs/readmes/README.validation.md]: thank you, updated.

File: docs/readmes/README.validation.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+`ConstraintTemplate` describes both the [Rego](https://www.openpolicyagent.org/docs/v0.10.7/how-do-i-write-policies/) that enforces the constraint and the schema of the constraint. 
+   
+   * User assigned MSI is expected to have Resource ID in the given format.
+
+   ```
+   /subscriptions/<subid>/resourcegroups/<resourcegroup>/providers/Microsoft.ManagedIdentity/userAssignedIdentities/<name>
+   ```
+
+   The same can be checked via following regex pattern, any resource not matching the pattern is considered as invalid.
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply docs/readmes/README.validation.md]: agree, rephrased as suggested.

File: docs/readmes/README.validation.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+kind: AzureIdentity
+metadata:
+  name: testidentityvalid
+spec:
+  type: 0
+  ResourceID: /subscriptions/00000000-0000-0000-0000-000000000000/resourcegroups/myResourceGroup/providers/Microsoft.ManagedIdentity/userAssignedIdentities/testidentity
+  ClientID: 00000000-0000-0000-0000-000000000000
+```
+
+   * Following identity will violate the constraint and request will be rejected,  as resource ID is not of correct format (`resourcegroups/<resourcegroup>` is missing in resourceID).
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply docs/readmes/README.validation.md]: done.

File: test/e2e/README.md [hunk 6]
```diff
@@ -114,6 +114,7 @@ To ensure consistency across all tests, they generally follow the format below:
 | Enable a user assigned identity on VMs, then assign a different user assigned identity to a pod | Pod identity should work as expected and the user assigned identity on VMs should not be altered after deleting the pod identity | Advanced |
 | Enable a user assigned identity on VMs, then assign the same user assigned identity to a pod | Pod identity should work as expected and the user assigned identity on VMs should not be altered after deleting the pod identity | Advanced |
 | Enable system assigned identity on VMs, then assign a user assigned identity to a pod | Pod identity should work as expected and the system assigned identity on VMs should not be altered after deleting the pod identity | Advanced |
+| Enforce user assinged identity format validation constraint via Gatekeeper | Azure identity with invalid resource ID should not be accepted | Advanced |
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply test/e2e/README.md]: fixed.

File: test/e2e/aadpodidentity_test.go [hunk 7]
```diff
@@ -916,6 +916,29 @@ var _ = Describe("Kubernetes cluster using aad-pod-identity", func() {
 
 		validateAzureAssignedIdentity(azureAssignedIdentity, keyvaultIdentity)
 	})
+
+	It("should pass the identity format validation via gatekeeper constraint", func() {
+
+		// setup the required infra
+		setupIdentityFormatValidationInfra()
+
+		// cleanup the infra craeted for this specific test
```
[the remark this answers is not in the document]
Comment (surenderssm) [reply test/e2e/aadpodidentity_test.go]: fixed.

**Comment by surenderssm** [comment]
Thank you @aramase , for your review and comments. I have acted on the comments.

**Comment by kkmsft** [comment]
Thank you, @surenderssm :-)

**Pull request merged**